AirCasting/SessionViews/ChartViewModel.swift

```diff
@@ -0,0 +1,96 @@
+// Created by Lunar on 30/09/2021.
+//
+
+import Foundation
+import Charts
+
+final class ChartViewModel: ObservableObject {
+    @Published var entries: [ChartDataEntry] = []
+    
+    private let stream: MeasurementStreamEntity
+
+    private var timeUnit: Double {
+        stream.session.type == .mobile ? 60 : 60 * 60
+    }
+    
+    private var mainTimer: Timer?
+    private var firstTimer: Timer?
+    
+    deinit {
+        mainTimer?.invalidate()
+        firstTimer?.invalidate()
+    }
+    
+    init(stream: MeasurementStreamEntity) {
+        self.stream = stream
+        generateEntries()
+        startTimers(stream.session)
+    }
+    
+    private func startTimers(_ session: SessionEntity) {
+        let timeOfNextAverage = timeOfNextAverage()
+
+        firstTimer = Timer.scheduledTimer(withTimeInterval: TimeInterval(timeOfNextAverage), repeats: false) { [weak self] timer in
+            self?.generateEntries()
+            self?.startMainTimer()
+        }
+    }
+    
+    func startMainTimer() {
+        mainTimer = Timer.scheduledTimer(withTimeInterval: timeUnit, repeats: true) { [weak self] timer in
+            self?.generateEntries()
+        }
+    }
+    
+    private func generateEntries() {
+        // Set up begning and end of the interval for the first average
+        //  - for fixed sessions we are taking the last full hour of the session, which has any measurements
```

Comment:
this is not accurate. For fixed sessions we should take the last 9 hours of the session. E.g. if session starts at 12:15 and now is 21:05, we'll have averages from:
12:15-12:59 (12)
13:00-13:59 (13)
...
20:00-20:59 (21)

Reply:
That's how it works (we are taking the last full hour). Maybe the wording in the comment is not accurate. Do you have an idea how to rephrase that? You can see in endTime function that we take `lastMeasurementTime.roundedDownToHour` as the end of the last interval (so in the example you hade it would be 21:05 rounded down to 21.